server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java

```diff
@@ -1040,8 +1043,98 @@ public void testDoNotRebalanceToTheNodeThatNoLongerExists() {
+        for (int n = 0; n < numberOfNodes; n++) {
+            discoveryNodeBuilder.add(newNode("node-" + n));
+        }
+
+        var assignments = new HashMap<ShardId, ShardAssignment>();
+        var indexRoutingTableBuilder = IndexRoutingTable.builder(index);
+
+        for (int i = 0; i < shardsPerNode * numberOfNodes; i++) {
+            var shardId = new ShardId(index, i);
+
+            var desiredNode = i % numberOfNodes;
+            var currentNode = randomIntBetween(0, numberOfNodes - 1);
+
+            assignments.put(shardId, new ShardAssignment(Set.of("node-" + desiredNode), 1, 0, 0));
+            indexRoutingTableBuilder.addShard(newShardRouting(shardId, "node-" + currentNode, true, STARTED));
+        }
+
+        var balance = new DesiredBalance(1, assignments);
+        var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(discoveryNodeBuilder)
+            .metadata(Metadata.builder().put(indexMetadata, true))
+            .routingTable(RoutingTable.builder().add(indexRoutingTableBuilder))
+            .build();
+
+        var clusterSettings = createBuiltInClusterSettings();
+        var deciders = new AllocationDecider[] {
+            new ConcurrentRebalanceAllocationDecider(clusterSettings),
+            new ThrottlingAllocationDecider(clusterSettings) };
+
+        var allocationOrdering = new NodeAllocationOrdering();
+        var moveOrdering = new NodeAllocationOrdering();
+
+        var totalOutgoingMoves = new HashMap<String, AtomicInteger>();
+        for (int i = 0; i < numberOfNodes; i++) {
+            var nodeId = "node-" + i;
+            if (isReconciled(clusterState.getRoutingNodes().node(nodeId), balance) == false) {
+                totalOutgoingMoves.put(nodeId, new AtomicInteger());
+            }
+        }
+
+        while (true) {
+
+            var allocation = createRoutingAllocationFrom(clusterState, deciders);
+            new DesiredBalanceReconciler(balance, allocation, allocationOrdering, moveOrdering).run();
+
+            var initializing = shardsWithState(allocation.routingNodes(), ShardRoutingState.INITIALIZING);
+            if (initializing.isEmpty()) {
+                break;
+            }
+            for (ShardRouting shardRouting : initializing) {
+                totalOutgoingMoves.get(shardRouting.relocatingNodeId()).incrementAndGet();
+                allocation.routingNodes().startShard(logger, shardRouting, allocation.changes(), 0L);
+            }
+
+            var summary = totalOutgoingMoves.values().stream().mapToInt(AtomicInteger::get).summaryStatistics();
+            // ensure that we do not cause hotspots by round-robin unreconciled source nodes when picking next rebalance
+            // (already reconciled nodes are excluded as they are no longer causing new moves)
+            assertThat(
+                "Every node expect to have similar amount of outgoing rebalances: " + totalOutgoingMoves,
```

Comment:
Sounds good, I updated the message to `Reconciling nodes should all have same amount (max 1 delta) of moves`

Reply:
I think that if we allowed 3 outgoing recoveries, we risked one node having just one shard left to move and another having 3 shards left to move. In that case, we'd have a delta of 2 even when we round-robin. Did I misunderstand something or does the check here not rely on max doing 2 outgoing recoveries at a time?

Reply:
If we allowed 3 outgoing recoveries I think we would risk one node having one outgoing relocation and the other node having 3 outgoing relocations, if these were the very last pending relocations needed. This would happen in one `DesiredBalanceReconciler.run` round, i.e., we would not remove from the `totalOutgoingMoves` before after incrementing the counts and validating - and I think that would make the test fail?

I should probably try this out....